File: platform_darwin.m
#import "platform_darwin.h"
#import "_cgo_export.h"

// ----- App section -----

@interface appDelegate : NSObject<NSApplicationDelegate>
@end

@implementation appDelegate

- (void)applicationWillFinishLaunching:(NSNotification *)aNotification {
	willFinishStartupCallback();
}

- (void)applicationDidFinishLaunching:(NSNotification *)aNotification {
	didFinishStartupCallback();
}

- (NSApplicationTerminateReply)applicationShouldTerminate:(NSApplication *)sender {
	// The Mac response codes map to the same values we use
	return checkQuitCallback();
}

- (BOOL)applicationShouldTerminateAfterLastWindowClosed:(NSApplication *)theApplication {
	return quitAfterLastWindowClosedCallback();
}

- (void)applicationWillTerminate:(NSNotification *)aNotification {
	quittingCallback();
}

- (void)applicationWillBecomeActive:(NSNotification *)aNotification {
	willActivateCallback();
}

- (void)applicationDidBecomeActive:(NSNotification *)aNotification {
	didActivateCallback();
}

- (void)applicationWillResignActive:(NSNotification *)aNotification {
	willDeactivateCallback();
}

- (void)applicationDidResignActive:(NSNotification *)aNotification {
	didDeactivateCallback();
}

@end

void prepareForStart() {
	[[NSAutoreleasePool alloc] init];
	[NSApplication sharedApplication];
	// Required for apps without bundle & Info.plist
	[NSApp setActivationPolicy:NSApplicationActivationPolicyRegular];
	[NSApp setDelegate:[appDelegate new]];
	// Required to use 'NSApplicationActivateIgnoringOtherApps' otherwise our windows end up in the background.
	[[NSRunningApplication currentApplication] activateWithOptions:NSApplicationActivateAllWindows | NSApplicationActivateIgnoringOtherApps];
}

void attemptQuit() {
	[NSApp terminate:nil];
}

void mayQuitNow(int quit) {
	[NSApp replyToApplicationShouldTerminate:quit];
}

void hideApp() {
	[NSApp hide:nil];
}

void hideOtherApps() {
	[NSApp hideOtherApplications:NSApp];
}

void showAllApps() {
	[NSApp unhideAllApplications:NSApp];
}

void invoke(unsigned long id) {
	dispatch_async_f(dispatch_get_main_queue(), (void *)id, (dispatch_function_t)dispatchUITaskCallback);
}

void invokeAfter(unsigned long id, long afterNanos) {
	dispatch_after_f(dispatch_time(DISPATCH_TIME_NOW, afterNanos), dispatch_get_main_queue(), (void *)id, (dispatch_function_t)dispatchUITaskCallback);
}

// ----- Menu section -----

@interface ItemDelegate : NSObject
@end

@implementation ItemDelegate

- (BOOL)validateMenuItem:(NSMenuItem *)menuItem {
	return validateMenuItemCallback(menuItem);
}

- (void)handleMenuItem:(id)sender {
	handleMenuItemCallback(sender);
}

@end

static ItemDelegate *itemDelegate = nil;

void setMenuBar(Menu bar) {
	[NSApp setMainMenu:bar];
}

void setServicesMenu(Menu menu) {
	[NSApp setServicesMenu:menu];
}

void setWindowMenu(Menu menu) {
	[NSApp setWindowsMenu:menu];
}

void setHelpMenu(Menu menu) {
	[NSApp setHelpMenu:menu];
}

Menu newMenu(const char *title) {
	return [[[NSMenu alloc] initWithTitle:[NSString stringWithUTF8String:title]] retain];
}

void disposeMenu(Menu menu) {
	[((NSMenu *)menu) release];
}

int menuItemCount(Menu menu) {
	return [((NSMenu *)menu) numberOfItems];
}

MenuItem menuItem(Menu menu, int index) {
	return (index < 0 || index >= menuItemCount(menu)) ?  nil : [((NSMenu *)menu) itemAtIndex:index];
}

void insertMenuItem(Menu menu, MenuItem item, int index) {
	[((NSMenu *)menu) insertItem:item atIndex:index];
}

void removeMenuItem(Menu menu, int index) {
	[((NSMenu *)menu) removeItemAtIndex:index];
}

MenuItem newMenuItem(const char *title, const char *selector, const char *key, int modifiers, bool needDelegate) {
	NSMenuItem *item = [[[NSMenuItem alloc] initWithTitle:[NSString stringWithUTF8String:title] action:NSSelectorFromString([NSString stringWithUTF8String:selector]) keyEquivalent:[NSString stringWithUTF8String:key]] retain];
	// macOS uses the same modifier mask bit order as we do, but it is shifted up by 16 bits
	[item setKeyEquivalentModifierMask:modifiers << 16];
	if (needDelegate) {
		if (!itemDelegate) {
			itemDelegate = [ItemDelegate new];
		}
		[item setTarget:itemDelegate];
	}
	return item;
}

Menu subMenu(MenuItem item) {
	NSMenuItem *mitem = (NSMenuItem *)item;
	if ([mitem hasSubmenu]) {
		return [mitem submenu];
	}
	return nil;
}

void setSubMenu(MenuItem item, Menu subMenu) {
	[((NSMenuItem *)item) setSubmenu: subMenu];
}

MenuItem newMenuSeparator() {
	return [[NSMenuItem separatorItem] retain];
}

void disposeMenuItem(MenuItem item) {
	[((NSMenuItem *)item) release];
}

// ----- Web view section -----

void add_ref_dummyx(cef_base_ref_counted_t *self) {}
int release_dummyx(cef_base_ref_counted_t *self) { return 1; }
int has_one_ref_dummyx(cef_base_ref_counted_t *self) { return 1; }
int has_at_least_one_ref_dummyx(cef_base_ref_counted_t *self) { return 1; }

void init_dummyx_ref_counter(cef_base_ref_counted_t *h) {
	h->add_ref = add_ref_dummyx;
	h->release = release_dummyx;
	h->has_one_ref = has_one_ref_dummyx;
	h->has_at_least_one_ref = has_at_least_one_ref_dummyx;
}

@interface EmbeddedWebView : NSView
{
	cef_browser_t *webview;
}
@end

@implementation EmbeddedWebView

- (instancetype)initWithFrame: (NSRect)frameRect url:(const char *)url {
	self = [super initWithFrame: frameRect];
	cef_window_info_t *info = (cef_window_info_t *)calloc(1, sizeof(cef_window_info_t));
	info->parent_view = self;

	cef_browser_settings_t *settings = (cef_browser_settings_t *)calloc(1, sizeof(cef_browser_settings_t));
	settings->size = sizeof(cef_browser_settings_t);

	cef_client_t *client = (cef_client_t *)calloc(1, sizeof(cef_client_t));
	client->base.size = sizeof(cef_client_t);
	init_dummyx_ref_counter(&client->base);
	cef_string_t *cefURL = (cef_string_t *)calloc(1, sizeof(cef_string_t));
	cef_string_from_utf8(url, strlen(url), cefURL);
	self->webview = cef_browser_host_create_browser_sync(info, client, cefURL, settings, nil);
	return self;
}

@end

// ----- Window section -----

@interface KeyableWindow : NSWindow
@end

@implementation KeyableWindow

-(BOOL)canBecomeKeyWindow {
	return YES;
}

@end

@interface WindowDelegate : NSObject<NSWindowDelegate>
@end

@implementation WindowDelegate

-(void)windowDidBecomeKey:(NSNotification *)notification {
	windowGainedKey((Window)[notification object]);
}

-(void)windowDidResignKey:(NSNotification *)notification {
	windowLostKey((Window)[notification object]);
}

-(BOOL)windowShouldClose:(id)sender {
	return (BOOL)windowShouldClose((Window)sender);
}

-(void)windowWillClose:(NSNotification *)notification {
	windowWillClose((Window)[notification object]);
}

@end

Window getKeyWindow() {
	return (Window)[NSApp keyWindow];
}

void bringAllWindowsToFront() {
	[[NSRunningApplication currentApplication] activateWithOptions:NSApplicationActivateAllWindows | NSApplicationActivateIgnoringOtherApps];
}

Window newWindow(int styleMask, double x, double y, double width, double height, const char *url) {
	// The styleMask bits match those that Mac OS uses
	NSRect contentRect = [NSWindow contentRectForFrameRect:NSMakeRect(x, [[NSScreen mainScreen] visibleFrame].size.height - (y + height), width, height) styleMask:styleMask];
	NSWindow *window = [[KeyableWindow alloc] initWithContentRect:contentRect styleMask:styleMask backing:NSBackingStoreBuffered defer:YES];
	EmbeddedWebView *view = [[EmbeddedWebView alloc] initWithFrame:NSMakeRect(0, 0, 0, 0) url:url];
	[window setContentView:view];
	[window setDelegate: [WindowDelegate new]];
	return (Window)window;
}

void closeWindow(Window window) {
	[((NSWindow *)window) close];
}

void setWindowTitle(Window window, const char *title) {
	[((NSWindow *)window) setTitle:[NSString stringWithUTF8String:title]];
}

void getWindowBounds(Window window, double *x, double *y, double *width, double *height) {
	CGRect frame = [((NSWindow *)window) frame];
	*x = frame.origin.x;
	*y = [[NSScreen mainScreen] visibleFrame].size.height - (frame.origin.y + frame.size.height);
	*width = frame.size.width;
	*height = frame.size.height;
}

void setWindowBounds(Window window, double x, double y, double width, double height) {
	NSWindow *win = (NSWindow *)window;
	CGRect frame = [win frame];
	[win setFrame:NSMakeRect(x, [[NSScreen mainScreen] visibleFrame].size.height - (y + height), width, height) display:YES];
}

void bringWindowToFront(Window window) {
	[((NSWindow *)window) makeKeyAndOrderFront:nil];
}

void minimizeWindow(Window window) {
	[((NSWindow *)window) performMiniaturize:nil];
}

void zoomWindow(Window window) {
	[((NSWindow *)window) performZoom:nil];
}

// ----- Displays section -----

Display *displays(unsigned long *qty) {
	[NSApplication sharedApplication];
	NSArray *s = [NSScreen screens];
	unsigned long count = [s count];
	*qty = count;
	Display *d = (Display *)malloc(sizeof(Display) * count);
	for (unsigned int i = 0; i < count; i++) {
		NSScreen *screen = [s objectAtIndex: i];
		d[i].scaleFactor = [screen backingScaleFactor];
		CGDirectDisplayID dID = (CGDirectDisplayID)[[[screen deviceDescription] objectForKey:@"NSScreenNumber"] unsignedIntValue];
		d[i].bounds = CGDisplayBounds(dID);
		d[i].usableBounds = [screen visibleFrame];
		d[i].isMain = CGDisplayIsMain(dID);
		CGRect b = [screen frame];
		d[i].usableBounds.origin.y = d[i].bounds.origin.y + (b.origin.y + b.size.height - (d[i].usableBounds.origin.y + d[i].usableBounds.size.height));
	}
	return d;
}

File: platform_darwin.go
package webapp

import (
	// #cgo CFLAGS: -x objective-c
	// #cgo LDFLAGS: -framework Cocoa -framework WebKit -framework Security
	// #include "platform_darwin.h"
	"C"
	"fmt"
	"strings"
	"time"
	"unsafe"

	"github.com/richardwilkes/toolbox/atexit"
	"github.com/richardwilkes/toolbox/cmdline"
	"github.com/richardwilkes/toolbox/i18n"
	"github.com/richardwilkes/toolbox/xmath/geom"
	"github.com/richardwilkes/webapp/keys"
)

// ----- App section -----

func platformStart() {
	C.start()
}

//export willFinishStartupCallback
func willFinishStartupCallback() {
	WillFinishStartupCallback()
}

//export didFinishStartupCallback
func didFinishStartupCallback() {
	DidFinishStartupCallback()
}

//export willActivateCallback
func willActivateCallback() {
	WillActivateCallback()
}

//export didActivateCallback
func didActivateCallback() {
	DidActivateCallback()
}

//export willDeactivateCallback
func willDeactivateCallback() {
	WillDeactivateCallback()
}

//export didDeactivateCallback
func didDeactivateCallback() {
	DidDeactivateCallback()
}

//export quitAfterLastWindowClosedCallback
func quitAfterLastWindowClosedCallback() bool {
	return QuitAfterLastWindowClosedCallback()
}

func platformAttemptQuit() {
	C.attemptQuit()
}

//export checkQuitCallback
func checkQuitCallback() int {
	return int(CheckQuitCallback())
}

func platformMayQuitNow(quit bool) {
	var mayQuit C.int
	if quit {
		mayQuit = 1
	}
	C.mayQuitNow(mayQuit)
}

//export quittingCallback
func quittingCallback() {
	QuittingCallback()
	atexit.Exit(0)
}

func platformInvoke(id uint64) {
	C.invoke(C.ulong(id))
}

func platformInvokeAfter(id uint64, after time.Duration) {
	C.invokeAfter(C.ulong(id), C.long(after.Nanoseconds()))
}

//export dispatchUITaskCallback
func dispatchUITaskCallback(id uint64) {
	dispatchUITask(id)
}

// ----- Menu section -----

var (
	appBar      *MenuBar
	menuMap     = make(map[C.Menu]*Menu)
	menuItemMap = make(map[C.MenuItem]*MenuItem)
)

type platformMenuBar struct {
}

type platformMenu struct {
	menu C.Menu
}

type platformMenuItem struct {
	item C.MenuItem
}

func platformMenuBarForWindow(_ *Window) *MenuBar {
	if appBar == nil {
		appBar = &MenuBar{
			platformMenuBar: platformMenuBar{}, // Here solely to shut the linter up
			bar:             NewMenu(""),
			special:         make(map[SpecialMenuType]*Menu),
			global:          true,
		}
		C.setMenuBar(appBar.bar.menu)
	}
	return appBar
}

func (bar *MenuBar) platformSetServicesMenu(menu *Menu) {
	C.setServicesMenu(menu.menu)
}

func (bar *MenuBar) platformSetWindowMenu(menu *Menu) {
	C.setWindowMenu(menu.menu)
}

func (bar *MenuBar) platformSetHelpMenu(menu *Menu) {
	C.setHelpMenu(menu.menu)
}

func (bar *MenuBar) platformFillAppMenu(appMenu *Menu) {
	appMenu.AppendItem(NewMenuSeparator())
	servicesMenu := NewMenu(i18n.Text("Services"))
	appMenu.AppendMenu(servicesMenu)
	bar.SetupSpecialMenu(ServicesSpecialMenu, servicesMenu)

	appMenu.AppendItem(NewMenuSeparator())
	hideItem := NewMenuItemWithKey(fmt.Sprintf(i18n.Text("Hide %s"), cmdline.AppName), keys.VirtualKeyH)
	hideItem.Handler = func() { C.hideApp() }
	appMenu.AppendItem(hideItem)
	hideOthersItem := NewMenuItemWithKeyAndModifiers(i18n.Text("Hide Others"), keys.VirtualKeyH, keys.OptionModifier|keys.PlatformMenuModifier())
	hideOthersItem.Handler = func() { C.hideOtherApps() }
	appMenu.AppendItem(hideOthersItem)
	showAllItem := NewMenuItem(i18n.Text("Show All"))
	showAllItem.Handler = func() { C.showAllApps() }
	appMenu.AppendItem(showAllItem)
}

func (menu *Menu) platformInit() {
	cTitle := C.CString(menu.title)
	menu.menu = C.newMenu(cTitle)
	menuMap[menu.menu] = menu
	C.free(unsafe.Pointer(cTitle))
}

func (menu *Menu) platformItemCount() int {
	return int(C.menuItemCount(menu.menu))
}

func (menu *Menu) platformItem(index int) *MenuItem {
	return menuItemMap[C.menuItem(menu.menu, C.int(index))]
}

func (menu *Menu) platformInsertItem(item *MenuItem, index int) {
	C.insertMenuItem(menu.menu, item.item, C.int(index))
}

func (menu *Menu) platformRemove(index int) {
	C.removeMenuItem(menu.menu, C.int(index))
}

func (menu *Menu) platformDispose() {
	C.disposeMenu(menu.menu)
	delete(menuMap, menu.menu)
}

func (item *MenuItem) platformInitMenuSeparator() {
	item.item = C.newMenuSeparator()
	menuItemMap[item.item] = item
}

func (item *MenuItem) platformInitMenuItem(kind MenuItemKind) {
	var keyCodeStr string
	if item.keyCode != 0 {
		mapping := keys.MappingForKeyCode(item.keyCode)
		if mapping.KeyChar != 0 {
			keyCodeStr = strings.ToLower(string(mapping.KeyChar))
		}
	}
	cTitle := C.CString(item.title)
	cKey := C.CString(keyCodeStr)
	cSelector := C.CString(kind.selector())
	item.item = C.newMenuItem(cTitle, cSelector, cKey, C.int(item.keyModifiers), kind == NormalKind)
	C.free(unsafe.Pointer(cSelector))
	C.free(unsafe.Pointer(cKey))
	C.free(unsafe.Pointer(cTitle))
	menuItemMap[item.item] = item
}

func (item *MenuItem) platformSubMenu() *Menu {
	if menu, ok := menuMap[C.subMenu(item.item)]; ok {
		return menu
	}
	return nil
}

func (item *MenuItem) platformSetSubMenu(subMenu *Menu) {
	C.setSubMenu(item.item, subMenu.menu)
}

func (item *MenuItem) platformDispose() {
	C.disposeMenuItem(item.item)
	delete(menuItemMap, item.item)
}

//export validateMenuItemCallback
func validateMenuItemCallback(menuItem C.MenuItem) bool {
	if item, ok := menuItemMap[menuItem]; ok && item.Validator != nil {
		item.enabled = item.Validator()
		return item.enabled
	}
	return true
}

//export handleMenuItemCallback
func handleMenuItemCallback(menuItem C.MenuItem) {
	if item, ok := menuItemMap[menuItem]; ok && item.Handler != nil {
		item.Handler()
	}
}

// ----- Window section -----

var (
	windowMap = make(map[C.Window]*Window)
)

type platformWindow struct {
	window C.Window
}

func platformBringAllWindowsToFront() {
	C.bringAllWindowsToFront()
}

func platformKeyWindow() *Window {
	if window, ok := windowMap[C.getKeyWindow()]; ok {
		return window
	}
	return nil
}

func (window *Window) platformInit(bounds geom.Rect, url string) {
	cURL := C.CString(url)
	window.window = C.newWindow(C.int(window.style), C.double(bounds.X), C.double(bounds.Y), C.double(bounds.Width), C.double(bounds.Height), cURL)
	C.free(unsafe.Pointer(cURL))
	windowMap[window.window] = window
}

func (window *Window) platformClose() {
	C.closeWindow(window.window)
	delete(windowMap, window.window)
}

func (window *Window) platformSetTitle(title string) {
	cTitle := C.CString(title)
	C.setWindowTitle(window.window, cTitle)
	C.free(unsafe.Pointer(cTitle))
}

func (window *Window) platformBounds() geom.Rect {
	var bounds geom.Rect
	C.getWindowBounds(window.window, (*C.double)(&bounds.X), (*C.double)(&bounds.Y), (*C.double)(&bounds.Width), (*C.double)(&bounds.Height))
	return bounds
}

func (window *Window) platformSetBounds(bounds geom.Rect) {
	C.setWindowBounds(window.window, C.double(bounds.X), C.double(bounds.Y), C.double(bounds.Width), C.double(bounds.Height))
}

func (window *Window) platformToFront() {
	C.bringWindowToFront(window.window)
}

func (window *Window) platformMinimize() {
	C.minimizeWindow(window.window)
}

func (window *Window) platformZoom() {
	C.zoomWindow(window.window)
}

//export windowGainedKey
func windowGainedKey(cWindow C.Window) {
	if window, ok := windowMap[cWindow]; ok {
		window.GainedFocus()
	}
}

//export windowLostKey
func windowLostKey(cWindow C.Window) {
	if window, ok := windowMap[cWindow]; ok {
		window.LostFocus()
	}
}

//export windowShouldClose
func windowShouldClose(cWindow C.Window) bool {
	if window, ok := windowMap[cWindow]; ok {
		return window.MayCloseCallback()
	}
	return true
}

//export windowWillClose
func windowWillClose(cWindow C.Window) {
	if window, ok := windowMap[cWindow]; ok {
		window.WillCloseCallback()
	}
}

// ----- Display section -----

func platformDisplays() []*Display {
	var count C.ulong
	ptr := unsafe.Pointer(C.displays(&count))
	displays := (*[1 << 30]C.Display)(ptr)
	result := make([]*Display, count)
	for i := range result {
		d := &Display{}
		d.Bounds.X = float64(displays[i].bounds.origin.x)
		d.Bounds.Y = float64(displays[i].bounds.origin.y)
		d.Bounds.Width = float64(displays[i].bounds.size.width)
		d.Bounds.Height = float64(displays[i].bounds.size.height)
		d.UsableBounds.X = float64(displays[i].usableBounds.origin.x)
		d.UsableBounds.Y = float64(displays[i].usableBounds.origin.y)
		d.UsableBounds.Width = float64(displays[i].usableBounds.size.width)
		d.UsableBounds.Height = float64(displays[i].usableBounds.size.height)
		d.ScaleFactor = float64(displays[i].scaleFactor)
		d.IsMain = displays[i].isMain != 0
		result[i] = d
	}
	C.free(ptr)
	return result
}
